### ISSUE TITLE
wine/wineserver bitness detection fails for wrapped Wine binaries

### ISSUE THREAD

**Issue by Matoking** [Winetricks/winetricks#2183]
In some environments such as Protontricks, Wine binaries are not directly available on $PATH. Instead, a Bash script is created for each Wine binary that takes care of setting up and running the underlying binary. These scripts are then made available through $PATH.

This breaks the "new WoW mode" detection introduced in bcd7a182de58c983b65db7f922f73e559bccaa37 that checks the architecture for `wine` and `wineserver`. This fails because the executables are no longer binaries as Winetricks expects, causing "Unknown file arch: 65" or other similar warnings.

This happens on Protontricks, Wine packaged on Nix/Nixpkgs and potentially other environments in which `wine` and/or `wineserver` uses a wrapper script.

Ideally, we should probably run a Wine command instead that we can use to determine whether `wine` has been compiled using new WoW mode. Or maybe there's some other way?

**Comment by austin987** [comment]
> In some environments such as Protontricks, Wine binaries are not directly available on $PATH. Instead, a Bash script is created for each Wine binary that takes care of setting up and running the underlying binary. These scripts are then made available through $PATH.
> 
> This breaks the "new WoW mode" detection introduced in [bcd7a18](https://github.com/Winetricks/winetricks/commit/bcd7a182de58c983b65db7f922f73e559bccaa37) that checks the architecture for `wine` and `wineserver`. This fails because the executables are no longer binaries as Winetricks expects, causing "Unknown file arch: 65" or other similar warnings.
> 
> This happens on Protontricks, Wine packaged on Nix/Nixpkgs and potentially other environments in which `wine` and/or `wineserver` uses a wrapper script.

Unfortunately there are a lot of wrappers out there, and each does things differently. E.g., #2035 has a patch for debian, but I suspect that would break things on macOS and possibly other platforms/wrappers.

I'd be happy to review a PR that can fix things in a generic way.

> Ideally, we should probably run a Wine command instead that we can use to determine whether `wine` has been compiled using new WoW mode. Or maybe there's some other way?

I'm not aware of any command that can determine that, but I'd be happy to be proven wrong.

**Comment by jre-wine** [comment]
https://github.com/Winetricks/winetricks/pull/2035 is not for wrapper scripts, but for a missing wine64 binary in path, which exists though in Wine's bindir (classic wow64)

https://github.com/Winetricks/winetricks/pull/2189 is for wrapper scripts. I'd be interested to hear if this helps for protontricks.

Both PRs should work just fine on all platforms, including MacOS.

**Comment by jre-wine** [comment]
Hm, couldn't protontricks set WINESERVER_BIN and WINE_BIN then?

Otherwise if we have to deal with scripts (and can't just completely get rid of the whole logic in favor of #2191), I suggest to just add all workarounds that might work, instead of trying to find one that works for all cases. I might come up with some workaround using WINEDLLPATH, but I'd like to hear first if this is needed at all.

Anyway, I changed #2189 again to only warn, but not exit. However there's an explicit warning now, that the WoW64 type could not be detected.

**Comment by Matoking** [comment]
> Hm, couldn't protontricks set WINESERVER_BIN and WINE_BIN then?
> 
> Otherwise if we have to deal with scripts (and can't just completely get rid of the whole logic in favor of #2191), I suggest to just add all workarounds that might work, instead of trying to find one that works for all cases. I might come up with some workaround using WINEDLLPATH, but I'd like to hear first if this is needed at all.
> 
> Anyway, I changed #2189 again to only warn, but not exit. However there's an explicit warning now, that the WoW64 type could not be detected.

Yes, Protontricks could set `WINESERVER_BIN` and `WINE_BIN` to point towards the underlying binaries. The current code in #2189 doesn't seem to support it, though it seems like an easy addition.

**Comment by rexendevar** [comment]
What am I supposed to do about this until this update is in a version of Winetricks I can download?

**Comment by austin987** [comment]
> What am I supposed to do about this until this update is in a version of Winetricks I can download?

What OS/distro/wine version are you using? A fix was merged for Debian; for protontricks/nix, I'd suggest filing a bug with their bug reporting trackers. If they are not using standard wine paths/binaries, they should either adapt their wrappers, or submit patches upstream to account for that.
